### PR TITLE
adding prometheus port configs for aggregating metrics in queue proxy

### PR DESCRIPTION
**Pull request by alexagriffith**
Signed-off-by: alexagriffith <agriffith50@bloomberg.net>

**What this PR does / why we need it**:

This PR follows https://github.com/kserve/kserve/pull/2425 where prometheus metrics were added in python sdk. 
This PR is the first of two that will allow the queue proxy to scrape metrics from the kserve-container. The reason this is needed is because of the limitations of knative only being able to scrape one port per pod. The end goal is to extend queue-proxy so that it scrapes metrics from kserve-container and aggregates them/exposes them on one port. A PoC for this already exists and. I am using it to test (this will be next PR). 

This PR's goal is to pass the kserve-container prometheus port into the queue-proxy so that queue-proxy can use it to scrape metrics and aggregate them. To do this, a new mutator is added. 

**Type of changes**
Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

**Feature/Issue validation/testing**:

Please describe the tests that you ran to verify your changes and relevent result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- [x] Unit test
- [x] Deployed each model and checked pod yaml and configs 

- Logs

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Checklist**:

- [x] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [x] Has code been commented, particularly in hard-to-understand areas?
- [x] Have you made corresponding changes to the documentation?

**Release note**:

```release-note

```

### CODE REVIEW

File: config/runtimes/kserve-tensorflow-serving.yaml [hunk 1]
```diff
@@ -3,6 +3,8 @@ kind: ClusterServingRuntime
 metadata:
   name: kserve-tensorflow-serving
 spec:
+  annotations:
```
Comment (alexagriffith) [inline config/runtimes/kserve-tensorflow-serving.yaml:6]: the only one I haven't figured out - docs are a little confusing. seems we should use --rest_api_port (8080 for us) https://www.tensorflow.org/tfx/serving/serving_config  but when I hit that with /metrics or /monitoring/prometheus/metrics , I get a `404`

File: pkg/webhook/admission/pod/metrics_aggregate_injector.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+	"github.com/kserve/kserve/pkg/constants"
+	v1 "k8s.io/api/core/v1"
+)
+
+const (
+	defaultPrometheusPort          = "8080"
+	PrometheusPortAnnotationKey    = "prometheus.io/port"
+	PrometheusPathAnnotationKey    = "prometheus.io/path"
+	PrometheusPathAnnotationValue  = "/metrics"
+	QueueProxyAggregateMetricsPort = "9088"
```
Comment (alexagriffith) [inline pkg/webhook/admission/pod/metrics_aggregate_injector.go]: I can move this to `constants` , if we think that would be a better place for them?

**Review by alexagriffith — COMMENTED** [review]
I need to add documentation about how to use 

```
annotations:
    serving.kserve.io/enable-metric-aggregation: "true"
    serving.kserve.io/set-prometheus-aggregate-annotation: "true"
```    

but with this PR, aggregation won't actually happen yet. It will in the next PR with the queue-proxy extension. Should I wait to add this documentation in the next PR? Also where is best to place this documentation?

File: config/runtimes/kserve-mlserver.yaml [hunk 3]
```diff
@@ -3,6 +3,8 @@ kind: ClusterServingRuntime
 metadata:
   name: kserve-mlserver
 spec:
+  annotations:
+    prometheus.kserve.io/port: '8080'
```
Comment (yuzisun) [inline config/runtimes/kserve-mlserver.yaml:8]: Does MLServer use port 8080 too ?

File: pkg/webhook/admission/pod/metrics_aggregate_injector_test.go [hunk 4]
```diff
@@ -0,0 +1,250 @@
+/*
+Copyright 2021 The KServe Authors.
```
Comment (yuzisun) [inline pkg/webhook/admission/pod/metrics_aggregate_injector_test.go]:
```suggestion
Copyright 2022 The KServe Authors.
```

File: pkg/webhook/admission/pod/metrics_aggregate_injector.go [hunk 5]
```diff
@@ -0,0 +1,61 @@
+/*
+Copyright 2021 The KServe Authors.
```
Comment (yuzisun) [inline pkg/webhook/admission/pod/metrics_aggregate_injector.go]:
```suggestion
Copyright 2022 The KServe Authors.
```

File: pkg/constants/constants.go [hunk 6]
```diff
@@ -176,9 +179,11 @@ const (
 
 // InferenceService Environment Variables
 const (
-	CustomSpecStorageUriEnvVarKey       = "STORAGE_URI"
-	CustomSpecProtocolEnvVarKey         = "PROTOCOL"
-	CustomSpecMultiModelServerEnvVarKey = "MULTI_MODEL_SERVER"
+	CustomSpecStorageUriEnvVarKey              = "STORAGE_URI"
+	CustomSpecProtocolEnvVarKey                = "PROTOCOL"
+	CustomSpecMultiModelServerEnvVarKey        = "MULTI_MODEL_SERVER"
+	KServeContainerPrometheusPortEnvVarKey     = "KSERVE_CONTAINER_PROMETHEUS_PORT"
```
Comment (yuzisun) [inline pkg/constants/constants.go]:
```suggestion
	KServeContainerPrometheusPortEnvVarKey     = "KSERVE_CONTAINER_METRICS_PORT"
```

Comment (yuzisun) [reply pkg/constants/constants.go]: Prometheus is one way of exposing metrics, might want to be agnostics of underlying technology here.

File: pkg/webhook/admission/pod/metrics_aggregate_injector.go [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+	PrometheusPathAnnotationValue  = "/metrics"
+	QueueProxyAggregateMetricsPort = "9088"
+)
+
+// InjectMetricsAggregator looks for the annotations to enable aggregate kserve-container and queue-proxy metrics and
+// if specified, sets port-related EnvVars in queue-proxy and the aggregate prometheus annotation.
+func InjectMetricsAggregator(pod *v1.Pod) error {
+	for i, container := range pod.Spec.Containers {
+		if container.Name == "queue-proxy" {
+			if enableMetricAgg, ok := pod.ObjectMeta.Annotations[constants.EnableMetricAggregation]; ok && enableMetricAgg == "true" {
```
Comment (yuzisun) [inline pkg/webhook/admission/pod/metrics_aggregate_injector.go:31]: I think we want a global configuration here in the [inference service configmap](https://github.com/kserve/kserve/blob/master/config/configmap/inferenceservice.yaml) instead of adding this annotation to each inference service. The annotation can be still used to override for the global value.

Comment (yuzisun) [reply pkg/webhook/admission/pod/metrics_aggregate_injector.go:31]: see https://istio.io/latest/docs/ops/integrations/prometheus/#option-1-metrics-merging

File: pkg/constants/constants.go (hunk 6 above)
Comment (alexagriffith) [reply pkg/constants/constants.go]: I think there are other metrics we can export though, do we want to be specific that this is just used for prometheus scraping?

File: pkg/webhook/admission/pod/metrics_aggregate_injector.go (hunk 7 above)
Comment (alexagriffith) [reply pkg/webhook/admission/pod/metrics_aggregate_injector.go:31]: the `enbableMetricAgg` annotation? I think here it checks if the annotation exists and is set to "true". I dont think we add the annotation to each inference service here right? I can add to configmag if needed, but just wondering how adding the enable metric agg annotation will help with this? we will still have this if statement right?

Comment (alexagriffith) [reply pkg/webhook/admission/pod/metrics_aggregate_injector.go:31]: the prometheus docs link to the prometheus annotation, do you mean that we should add that one to configmap?

File: config/runtimes/kserve-mlserver.yaml (hunk 3 above)
Comment (alexagriffith) [reply config/runtimes/kserve-mlserver.yaml:8]: good catch, I think I lumped it with the python ones. I think its 8082, checking.

Comment (alexagriffith) [reply config/runtimes/kserve-mlserver.yaml:8]: mlserver docs seems to suggest its using port 8082 https://mlserver.readthedocs.io/en/latest/reference/settings.html

but I tested it and 8082 doesn't emit prometheus metrics, but 8080/metrics does.

Comment (alexagriffith) [reply config/runtimes/kserve-mlserver.yaml:8]: tested using https://github.com/kserve/kserve/tree/master/docs/samples/v1beta1/xgboost and confirmed mlserver was used because it set 8082 as the env var

File: pkg/webhook/admission/pod/metrics_aggregate_injector.go (hunk 7 above)
Comment (alexagriffith) [reply pkg/webhook/admission/pod/metrics_aggregate_injector.go:31]: right now default is false (or no annotation set), are you suggesting we change it to true? the only issue with that I think is the custom queue-proxy image. maybe we want to leave default as not present / false for now.

File: pkg/constants/constants.go (hunk 6 above)
Comment (yuzisun) [reply pkg/constants/constants.go]: I think the name is a bit confusing, the port is for the metrics endpoint and prometheus is really the metrics format.

File: pkg/webhook/admission/pod/metrics_aggregate_injector.go (hunk 7 above)
Comment (yuzisun) [reply pkg/webhook/admission/pod/metrics_aggregate_injector.go:31]: We should add to the configmap but still default to false, when user configures to be true then we enable the metrics merging in the mutator and set the env on queue proxy. Without adding to configmap user needs to add the `EnableMetricAggregation` annotation to every inference service yaml if we want to turn on this feature across board right?

When configmap is set to false, the if condition can be still kept to override the configmap default to enable the metrics merging if you want to turn on selective inference service.

File: pkg/constants/constants.go [hunk 8]
```diff
@@ -77,6 +77,9 @@ var (
 	MinScaleAnnotationKey                       = KnativeAutoscalingAPIGroupName + "/minScale"
 	MaxScaleAnnotationKey                       = KnativeAutoscalingAPIGroupName + "/maxScale"
 	RollOutDurationAnnotationKey                = KnativeServingAPIGroupName + "/rollout-duration"
+	EnableMetricAggregation                     = KServeAPIGroupName + "/enable-metric-aggregation"
+	KserveContainerPrometheusPort               = "prometheus.kserve.io/port"
+	SetPrometheusAggregateAnnotation            = KServeAPIGroupName + "/set-prometheus-aggregate-annotation"
```
Comment (yuzisun) [inline pkg/constants/constants.go]: I think we can name it as `enable-prometheus-scraping`, wdyt?

File: pkg/webhook/admission/pod/metrics_aggregate_injector.go (hunk 7 above)
Comment (yuzisun) [reply pkg/webhook/admission/pod/metrics_aggregate_injector.go:31]: We can add this in a separate PR though.

Comment (alexagriffith) [reply pkg/webhook/admission/pod/metrics_aggregate_injector.go:31]: So right now if the annotation is not present or it is set to false explicitly (not likely currently) it will not set the aggregation annotations. So at this point the annotation is only added to yaml if the inference service specifies it.

I see what you are saying, I can set a default to false, and every pod yaml will have this annotation where at the moment some pod yaml may not have it at all.

Comment (alexagriffith) [reply pkg/webhook/admission/pod/metrics_aggregate_injector.go:31]: Or *it won't set the aggregation env vars/ annotation

File: pkg/constants/constants.go (hunk 8 above)
Comment (alexagriffith) [reply pkg/constants/constants.go]: Enable Prometheus scraping may be confusing because there is an annotation Prometheus.io/scrape: true that can be set and this is separate from that

Comment (yuzisun) [reply pkg/constants/constants.go]: Though this annotation does not sound like something for user to configure and worry about, we might want to avoid extra yaml UX burden for people.

File: config/runtimes/kserve-tensorflow-serving.yaml (hunk 1 above)
Comment (yuzisun) [reply config/runtimes/kserve-tensorflow-serving.yaml:6]: Did you try 8501? See https://stackoverflow.com/questions/52975301/prometheus-for-tensorflow-serving

File: config/runtimes/kserve-mlserver.yaml (hunk 3 above)
Comment (yuzisun) [reply config/runtimes/kserve-mlserver.yaml:8]: From mlserver maintainer, the longer version is that 1.0.0 exposed the metrics endpoint in the same HTTP server as the REST inference endpoints. But on 1.1.0 we decided to move that to a separate server (which runs on port 8082 by default).

Comment (yuzisun) [reply config/runtimes/kserve-mlserver.yaml:8]: From mlserver maintainer, the longer version is that 1.0.0 exposed the metrics endpoint in the same HTTP server as the REST inference endpoints. But on 1.1.0 we decided to move that to a separate server (which runs on port 8082 by default).

Comment (alexagriffith) [reply config/runtimes/kserve-mlserver.yaml:8]: Okay, so we are using earlier version?

File: config/runtimes/kserve-tensorflow-serving.yaml (hunk 1 above)
Comment (alexagriffith) [reply config/runtimes/kserve-tensorflow-serving.yaml:6]: yes

File: pkg/webhook/admission/pod/metrics_aggregate_injector.go (hunk 7 above)
Comment (alexagriffith) [reply pkg/webhook/admission/pod/metrics_aggregate_injector.go:31]: after discussing online: in the next PR this will be set in config map at a global level to give infra engineers the ability to set a global default for their services. since it doesn't actually scrape metrics until the next PR, it will be added then.

File: config/runtimes/kserve-mlserver.yaml (hunk 3 above)
Comment (alexagriffith) [reply config/runtimes/kserve-mlserver.yaml:8]: added a comment in the runtime YAML about this.

File: pkg/constants/constants.go (hunk 8 above)
Comment (alexagriffith) [reply pkg/constants/constants.go]: changing to `enable-prometheus-aggregate-scraping`

File: pkg/constants/constants.go (hunk 6 above)
Comment (alexagriffith) [reply pkg/constants/constants.go]: changed
